[MXNET-1083] Add the example to demonstrate the inference workflow using C++ API

leleamol opened the pull request:
## Description ##
The PR includes an example that demonstrates the inference workflow using C++ API.


## Checklist ##
### Essentials ###
Please feel free to remove inapplicable items for your PR.
- [y] The PR title starts with [MXNET-$JIRA_ID], where $JIRA_ID refers to the relevant [JIRA issue](https://issues.apache.org/jira/projects/MXNET/issues) created (except PRs with tiny changes)
- [ ] Changes are complete (i.e. I finished coding on this PR)
- [ ] All changes have test coverage:
- Unit tests are added for small changes to verify correctness (e.g. adding a new operator)
- Nightly tests are added for complicated/long-running ones (e.g. changing distributed kvstore)
- Build tests will be added for build configuration changes (e.g. adding a new build option with NCCL)
- [ ] Code is well-documented: 
- For user-facing API changes, API doc string has been updated. 
- For new C++ functions in header files, their functionalities and arguments are documented. 
- For new examples, README.md is added to explain the what the example does, the source of the dataset, expected performance on test set and reference to the original paper if applicable
- Check the API doc at http://mxnet-ci-doc.s3-accelerate.dualstack.amazonaws.com/PR-$PR_ID/$BUILD_ID/index.html
- [ ] To the my best knowledge, examples are either not affected by this change, or have been fixed to be compatible with this change

### Changes ###
- [ ] Feature1, tests, (and when applicable, API doc)
- [ ] Feature2, tests, (and when applicable, API doc)
The example demonstrates how to load the pre-trained model and associated parameter files.
The model and parameter files can be specified as command line arguments.

## Comments ##
- If this change is a backward incompatible change, why must this change be made.
- Interesting edge cases to note here

leleamol commented:
@nswamy addressed the latest review comments
1. Loading the mean image only once.
2. Remove the image_data instance variable.

lebeg commented on cpp-package/example/inference/Makefile:
```diff
@@ -0,0 +1,43 @@
+# under the License.
+
+
+prebuild :
```
Would it be possible to make the example build in CMake instead? The handling of the flags would me much easier as a bonus.

leleamol replied:
Yes. But I would like to defer it to a later PR when we have more examples in this directory.

lebeg commented on cpp-package/example/inference/Makefile:
```diff
@@ -0,0 +1,43 @@
+
+prebuild :
+	$(shell cp -r ../../../lib ./)
+	$(shell wget -nc -O mean_224.nd https://github.com/dmlc/web-data/raw/master/mxnet/example/feature_extract/mean_224.nd)
```
It would be better to separate the build and dependency installation.

leleamol replied:
I am keeping the same style as that of other examples in cpp-package/examples directory.

leleamol replied:
Fixed.

lebeg commented on cpp-package/example/inference/inception_inference.cpp:
```diff
@@ -0,0 +1,391 @@
+    map<string, NDArray> aux_map;
+    std::vector<std::string> output_labels;
+    Symbol net;
+    Executor *executor;
```
Maybe use an owning `std::unique_ptr`?

leleamol replied:
The underlying code is returning a pointer and I couldnot use std::make_unique to convert it into unique_ptr

lebeg replied:
You don't need to create a `std::unique_ptr` via `std::make_unique`, you can just use the [constructor](https://en.cppreference.com/w/cpp/memory/unique_ptr/unique_ptr) with the given raw pointer as argument.

lebeg commented on cpp-package/example/inference/inception_inference.cpp:
```diff
@@ -0,0 +1,391 @@
+ */
+void Predictor::LoadModel(const std::string& model_json_file) {
+  LG << "Loading the model from " << model_json_file << std::endl;
+  net = Symbol::Load(model_json_file);
```
I suggest to add some use friendly error messages for every step. For example:
```
Model file not found
```

```
Parameter file not found
```

leleamol replied:
Will do.

leleamol replied:
Fixed.

lebeg commented on cpp-package/example/inference/inception_inference.cpp:
```diff
@@ -0,0 +1,391 @@
+  NDArray image_data = LoadInputImage(image_file);
+
+  // Normalize the image
+  if (!mean_image_file.empty()) {
```
What happens if it is?

leleamol replied:
Will add the explicit comment. If the mean file is not present it affects the accuracy slightly.

leleamol replied:
Fixed.

lebeg commented on cpp-package/example/inference/inception_inference.cpp:
```diff
@@ -0,0 +1,391 @@
+  auto array = executor->outputs[0].Copy(global_ctx);
+  NDArray::WaitAll();
+
+  float best_accuracy = 0.0;
```
It's better to use the literal in this case: `0.f` or make `best_accuracy` double.

leleamol replied:
Fixed.

lebeg commented on cpp-package/example/inference/inception_inference.cpp:
```diff
@@ -0,0 +1,391 @@
+    int num_unit = strtol(hidden_units_string.c_str(), &pNext, 10);
+    dimensions.push_back(num_unit);
+    while (*pNext) {
+        num_unit = strtol(pNext, &pNext, 10);
```
[`std::stringstream`](https://en.cppreference.com/w/cpp/io/basic_stringstream) with [`>>`](https://en.cppreference.com/w/cpp/io/basic_istream/operator_gtgt) operators could be much more readable and safe.

leleamol replied:
Can you elaborate a little more?

lebeg replied:
```cpp
std::istringstream stream(hidden_units_string);
std::vector<index_t> dimensions;
int num_unit;
while (stream >> num_unit) {
  dimensions.push_back(num_unit);
}
```

leleamol replied:
@lebeg The "strtol method" was recommended by @larroy in my last PR, hence I am using is to keep it consistent.

lebeg replied:
@larroy which method would you suggest?

kalyc commented:
@mxnet-label-bot add [pr-awaiting-review]

stu1130 commented on cpp-package/example/inference/unit_test_inception_inference.sh:
```diff
@@ -0,0 +1,39 @@
+else
+    echo "FAIL: inception_inference FAILED to identify the image."
+    exit 1
+fi
```
I think we can put all the training and inference unit tests under the unittest folder? and why not use C++ testing framework such as [Catch](https://github.com/catchorg/Catch2)

lebeg replied:
Catch is 👍 but gtest is already available in MXNet as a subrepo, so it might be considered as well

leleamol replied:
The examples are written so that they won't have external dependency.  However, will definitely consider gtest for unit tests.

stu1130 commented:
@mxnet-label-bot update [pr-awaiting-response]

vandanavk commented:
Could you add information about this example in the README file `cpp-package/example/README.md`

leleamol commented on cpp-package/example/inference/Makefile:
```diff
@@ -0,0 +1,43 @@
+
+
+prebuild :
+	$(shell cp -r ../../../lib ./)
```
Fixed.

leleamol commented:
@vandanavk  Added the ReadMe file for example.

lebeg commented on cpp-package/example/inference/inception_inference.cpp:
```diff
@@ -0,0 +1,414 @@
+        return (stat(name.c_str(), &buffer) == 0);
+    }
+    NDArray mean_img;
+    map<string, NDArray> args_map;
```
nit: Referencing `std` classes seem inconsistent throughout the file. Maybe keeping `std::` everywhere and removing `using namespace std;` could be an option.

leleamol replied:
Fixed.

lebeg commented on cpp-package/example/inference/inception_inference.cpp:
```diff
@@ -0,0 +1,414 @@
+    char *pNext;
+    int num_unit = strtol(hidden_units_string.c_str(), &pNext, 10);
+    dimensions.push_back(num_unit);
+    while (*pNext) {
```
In case you would keep this method the naming of the `pNext` variable differs from the convention (lower case with underscore).

leleamol replied:
Fixed.
IMO, usage of strtol is good since this is an example to demonstrate inference.

leleamol replied:
@lebeg , What I meant was for the purpose of this example usage of strtol can be acceptable.
Thanks for restarting the build.

lebeg commented on cpp-package/example/inference/inception_inference.cpp:
```diff
@@ -0,0 +1,414 @@
+              << "--synset file containing labels for prediction "
+              << "[--input_shape <dimensions of input image e.g \"3 224 224\"]"
+              << "[--mean file containing mean image for normalizing the input image "
+              << "[--gpu]  Specify this option if workflow needs to be run in gpu context "
```
Formatting: The `]` bracket is not consistent with other options. An extra whitespace between the option and the description.

leleamol replied:
Fixed.

lebeg commented on cpp-package/example/inference/inception_inference.cpp:
```diff
@@ -0,0 +1,414 @@
+            index++;
+            model_file_json = argv[index];
+        } else if (strcmp("--params", argv[index]) == 0) {
+            index++;
```
You need to check bounds here as well or an empty parameter can lead to a crash.

leleamol replied:
Fixed.

lebeg commented on cpp-package/example/inference/inception_inference.cpp:
```diff
@@ -0,0 +1,414 @@
+    // Run the forward pass to predict the image.
+    predict.PredictImage(input_image);
+  } catch (std::runtime_error &error) {
+    LG << "Execution failed with ERROR: " << error.what();
```
Maybe 2 exceptional cases can be combined together? If exception is `std::runtime_error` it's not an MXNet error?

leleamol replied:
I tested by giving wrong inputs such as incorrect input dimensions or corrupted json file.  It's not obvious which type of exception is thrown by MXNet engine. It was a generic exception.

leleamol commented:
@mxnet-jenkins  The recent CI failure doesn't seem to be related to my changes.  Can we re trigger the build?

lebeg commented:
Job timed out due to dockerhub failure. Restarted.

leleamol commented:
Can this PR be merged? I  have addressed the review comments.  The example is working as expected and demonstrates the usage of C++ API in running inference workflow.

lupesko commented:
@mxnet-label-bot update [pr-awaiting-review]
@sandeep-krishnamurthy @nswamy @anirudh2290 - can you guys review/merge please?

sandeep-krishnamurthy commented on cpp-package/example/inference/inception_inference.cpp:
```diff
@@ -0,0 +1,414 @@
+  LG << "Load the mean image data that will be used to normalize "
+     << "the image before running forward pass.";
+  mean_image_data = NDArray(input_shape, global_ctx, false);
+  mean_image_data.SyncCopyFromCPU(
```
If user does not provide one, should we use a default normalization - https://github.com/dmlc/gluon-cv/blob/master/gluoncv/utils/export_helper.py#L14

leleamol replied:
Done. I have used the mean values from the example.
I could not use the broadcast operators (such as broadcast_sub or broadcast_min) because they are not yet supported (in a user friendly way) for NDArray in C++ bindings.

We can use them later once they are implemented in NDArray.

sandeep-krishnamurthy commented on cpp-package/example/inference/inception_inference.cpp:
```diff
@@ -0,0 +1,414 @@
+  std::size_t best_idx = 0;
+
+  // Find out the maximum accuracy and the index associated with that accuracy.
+  for (std::size_t i = 0; i < array.Size(); ++i) {
```
may be use NDArray's argmax?

leleamol replied:
Done.

sandeep-krishnamurthy commented on cpp-package/example/inference/inception_inference.cpp:
```diff
@@ -0,0 +1,414 @@
+    return 1;
+  }
+
+  std::vector<index_t> input_dimensions = getShapeDimensions(input_shape);
```
nit: why not use tuple?

leleamol replied:
The Shape() constructor accepts a std::vector hence used the vector.

leleamol commented on cpp-package/example/inference/README.md:
````diff
@@ -0,0 +1,35 @@
+	```
+	./inception_inference --help
+	Usage:
+	inception_inference --symbol <model symbol file in json format>  --params <model params file> --image <path to the image used for prediction --synset file containing labels for prediction [--input_shape <dimensions of input image e.g "3 224 224"][--mean file containing mean image for normalizing the input image [--gpu]  Specify this option if workflow needs to be run in gpu context
````
Re-organized the comments.
The mean image file contains the mean values in image format. The file is also used in other feature extract example.
I have added the code to use the default mean values if user does not specify this file.

leleamol commented on cpp-package/example/inference/inception_inference.cpp:
```diff
@@ -0,0 +1,414 @@
+/*
+ * This example demonstrates image classification workflow with pre-trained models using MXNet C++ API.
+ * The example performs following tasks.
+ * 1. Load the pre-trained model,
```
Done

leleamol commented on cpp-package/example/inference/inception_inference.cpp:
```diff
@@ -0,0 +1,414 @@
+ public:
+    Predictor() {}
+    Predictor(const std::string& model_json,
+              const std::string& model_params,
```
Done.

leleamol commented on cpp-package/example/inference/inception_inference.cpp:
```diff
@@ -0,0 +1,414 @@
+  }
+  LG << "Loading the model parameters from " << model_parameters_file << std::endl;
+  std::map<std::string, NDArray> paramters;
+  NDArray::Load(model_parameters_file, 0, &paramters);
```
Done.

leleamol commented on cpp-package/example/inference/inception_inference.cpp:
```diff
@@ -0,0 +1,414 @@
+  std::ifstream fi(synset_file.c_str());
+  if (!fi.is_open()) {
+    std::cerr << "Error opening synset file " << synset_file << std::endl;
+    assert(false);
```
This is actually redundant. I have changed assert to throw an exception.

leleamol commented on cpp-package/example/inference/inception_inference.cpp:
```diff
@@ -0,0 +1,414 @@
+  }
+  LG << "Loading the model parameters from " << model_parameters_file << std::endl;
+  std::map<std::string, NDArray> paramters;
+  NDArray::Load(model_parameters_file, 0, &paramters);
```
The '0' indicates a null pointer and not the epoch number.

leleamol commented on cpp-package/example/inference/inception_inference.cpp:
```diff
@@ -0,0 +1,414 @@
+              const std::string& synset_file = "",
+              const std::string& mean_image_file = "");
+    void PredictImage(const std::string& image_file);
+    ~Predictor();
```
It is done in destructor.

leleamol commented on cpp-package/example/inference/README.md:
```diff
@@ -0,0 +1,35 @@
+
+## Examples demonstrating inference workflow
+
+This directory contains following examples. In order to run the examples, ensure that the path to the MXNet shared library is added to the OS specific environment variable viz. **LD\_LIBRARY\_PATH** for Linux, Mac and Ubuntu OS and **PATH** for Windows OS.
```
The example command is included in the README below.

leleamol commented on cpp-package/example/README.md:
```diff
@@ -2,7 +2,8 @@
 ## Building C++ examples
 
-The examples are built while building the MXNet library and cpp-package from source . However, they can be built manually as follows
+The examples in this folder demonstrate the **training** workflow. The **inference workflow** related examples can be found in [inference](<https://github.com/apache/incubator-mxnet/blob/master/cpp-package/example/inference>) folder.
```
Agree.

leleamol commented:
I have addressed all the review comments. Can this be merged? Since the purpose of this example is to show the inference workflow, I would like  to keep the example simple.

leleamol commented:
@mxnet-label-bot remove [pr-awaiting-merge]
